pallets/parachain-staking/src/lib.rs

```diff
@@ -389,287 +449,705 @@ pub mod pallet {
-			);
-			let mut delegation_amt: Option<B> = None;
-			self.top_delegations = self
-				.top_delegations
+			candidate: &T::AccountId,
+			delegator: T::AccountId,
+		) -> Result<bool, DispatchError>
+		where
+			BalanceOf<T>: Into<Balance>,
+		{
+			let old_total_counted = self.total_counted;
+			// remove top delegation
+			let mut top_delegations =
+				<TopDelegations<T>>::get(candidate).expect("existence proof TODO");
+			let mut actual_amount_option: Option<BalanceOf<T>> = None;
+			top_delegations.delegations = top_delegations
+				.delegations
 				.clone()
 				.into_iter()
 				.filter_map(|d| {
 					if d.owner != delegator {
 						Some(d)
 					} else {
-						delegation_amt = Some(d.amount);
+						actual_amount_option = Some(d.amount);
 						None
 					}
 				})
 				.collect();
-			// item removed from the top => highest bottom is popped from bottom and pushed to top
-			if let Some(amount) = delegation_amt {
-				// last element has largest amount as per ordering
-				if let Some(last) = self.bottom_delegations.pop() {
-					self.total_counted -= amount - last.amount;
-					self.add_top_delegation(last);
-				} else {
-					// no item in bottom delegations so no item from bottom to pop and push up
-					self.total_counted -= amount;
-				}
-				self.total_backing -= amount;
-				return Ok((true, amount));
+			let actual_amount = actual_amount_option.ok_or(Error::<T>::DelegationDNE)?;
+			top_delegations.total -= actual_amount;
+			// if bottom nonempty => bump top bottom to top
+			if !matches!(self.bottom_capacity, CapacityStatus::Empty) {
+				let mut bottom_delegations = <BottomDelegations<T>>::get(candidate)
+					.expect("bottom is nonempty as just checked");
+				// expect already stored greatest to least by bond amount
+				let lowest_bottom_delegation = bottom_delegations.delegations.pop().expect("");
+				bottom_delegations.total -= lowest_bottom_delegation.amount;
+				self.reset_bottom_data::<T>(&bottom_delegations);
+				<BottomDelegations<T>>::insert(candidate, bottom_delegations);
+				// insert lowest bottom into top delegations
+				top_delegations.insert_sorted_greatest_to_least(lowest_bottom_delegation);
 			}
-			// else (no item removed from the top)
-			self.bottom_delegations = self
-				.bottom_delegations
+			// update candidate info
+			self.reset_top_data::<T>(&top_delegations);
+			self.delegation_count -= 1u32;
+			<TopDelegations<T>>::insert(candidate, top_delegations);
+			// return whether total counted changed
+			Ok(old_total_counted == self.total_counted)
+		}
+		/// Remove bottom delegation
+		/// Returns if_total_counted_changed: bool
+		pub fn rm_bottom_delegation<T: Config>(
+			&mut self,
+			candidate: &T::AccountId,
+			delegator: T::AccountId,
+		) -> Result<bool, DispatchError>
+		where
+			BalanceOf<T>: Into<Balance>,
+		{
+			// remove bottom delegation
+			let mut bottom_delegations =
+				<BottomDelegations<T>>::get(candidate).expect("existence proof TODO");
```

Comment:
Is this TODO relevant still?

Reply:
I'll write the proof, it is simply the CandidateInfo exists => BottomDelegations exists